lib/benchee/benchmark/runner.ex

```diff
@@ -0,0 +1,154 @@
+defmodule Benchee.Benchmark.Runner do
+  @moduledoc """
+  This module actually runs our benchmark scenarios, adding information about
+  run time and memory usage to each scenario.
+  """
+
+  alias Benchee.Benchmark
+  alias Benchee.Benchmark.{Scenario, ScenarioContext}
+  alias Benchee.Utility.RepeatN
+  alias Benchee.Configuration
+
+  @doc """
+  Executes the benchmarks defined before by first running the defined functions
+  for `warmup` time without gathering results and them running them for `time`
+  gathering their run times.
+
+  This means the total run time of a single benchmarking job is warmup + time.
+
+  Warmup is usually important for run times with JIT but it seems to have some
+  effect on the BEAM as well.
+
+  There will be `parallel` processes spawned executing the benchmark job in
+  parallel.
+  """
+  @spec run_scenarios([Scenario.t], ScenarioContext.t) :: [Scenario.t]
+  def run_scenarios(scenarios, scenario_context) do
+    Enum.flat_map(scenarios, fn(scenario) ->
+      parallel_benchmark(scenario, scenario_context)
+    end)
+  end
+
+  defp parallel_benchmark(scenario = %Scenario{job_name: job_name, input_name: input_name},
+                          scenario_context = %ScenarioContext{printer: printer, config: config}) do
+    printer.input_information(input_name, config)
+    printer.benchmarking(job_name, config)
+    pmap 1..config.parallel, fn ->
+      run_warmup(scenario, scenario_context)
+      run_benchmark(scenario, scenario_context)
+    end
+  end
+
+  def run_warmup(scenario, scenario_context = %ScenarioContext{
+                   config: %Configuration{warmup: warmup}
+                 }) do
+    warmup_context = %ScenarioContext{scenario_context | show_fast_warning: false,
+                                                         run_time: warmup}
```

Comment:
Hm yeah not a huge of duplicating these keys alongside having them in config, first thought was "wait... can you do this, this has to be on config, doesn't it?" More comments on the struct :)